pypose/module/pnp.py

```diff
@@ -0,0 +1,493 @@
+            >>> pose = pypose.SE3([ 0.0000, -8.0000,  0.0000,  0.0000, -0.3827,  0.0000,  0.9239])
+            >>> f = 2
+            >>> img_size = (7, 7)
+            >>> projection_matrix = torch.tensor([[f, 0, img_size[0] / 2,], [0, f, img_size[1] / 2,], [0, 0, 1, ]])
+            >>> # some random points in the view
+            >>> pts_c = torch.tensor([[2., 0., 2.], [1., 0., 2.], [0., 1., 1.], [0., 0., 1.], [5., 5., 3.]])
+            >>> pixels = (pts_c @ projection_matrix.T)[:, :2] / (pts_c @ projection_matrix.T)[:, 2:]
+            >>> pixels
+            tensor([[5.5000, 3.5000],
+                    [4.5000, 3.5000],
+                    [3.5000, 5.5000],
+                    [3.5000, 3.5000],
+                    [6.8333, 6.8333]])
+            >>> # transform the points to world coordinate
+            >>> # solve the PnP problem to find the camera pose
+            >>> pts_w = pose.Inv().Act(pts_c)
+            >>> # solve the PnP problem
+            >>> epnp = pypose.module.EPnP()
+            >>> # when input is not batched, remember to add a batch dimension
+            >>> pose = epnp(pts_w[None], pixels[None], projection_matrix[None])
+            >>> pose
+            SE3Type LieTensor:
+            LieTensor([[ 5.4955e-05, -8.0000e+00, -2.7895e-05,  6.8488e-06, -3.8270e-01,
+                         3.2812e-06,  9.2387e-01]])
+
+    """
+
+    class BetasOptimizationObjective(torch.nn.Module):
+        """
+        Optimize the betas according to the objectives in the paper.
+        For the details, please refer to equation 15.
+        """
+
+        def __init__(self, betas):
+            super().__init__()
+            self.betas = torch.nn.Parameter(betas)
+
+        def forward(self, ctrl_pts_w, kernel_bases):
+            """
+            Args:
+                ctrl_pts_w: The control points in world coordinate. The shape is (B, 4, 3).
+                kernel_bases: The kernel bases. The shape is (B, 16, 4).
+            Returns:
+                torch.Tensor: The loss. The shape is (B, ).
+            """
+            batch_shape = kernel_bases.shape[:-2]
+            # calculate the control points in camera coordinate
+            ctrl_pts_c = pypose.bmv(kernel_bases, self.betas)
+            diff_c = ctrl_pts_c.reshape(*batch_shape, 1, 4, 3) - ctrl_pts_c.reshape(*batch_shape, 4, 1, 3)
+            diff_c = diff_c.reshape(*batch_shape, 16, 3)
+            diff_c = torch.sum(diff_c ** 2, dim=-1)
+
+            # calculate the distance between control points in world coordinate
+            diff_w = ctrl_pts_w.reshape(*batch_shape, 1, 4, 3) - ctrl_pts_w.reshape(*batch_shape, 4, 1, 3)
+            diff_w = diff_w.reshape(*batch_shape, 16, 3)
+            diff_w = torch.sum(diff_w ** 2, dim=-1)
+
+            error = (diff_w - diff_c).abs().mean(dim=-1)
+
+            return error
+
+    def __init__(self, naive_ctrl_pts=False, refinement_optimizer=GaussNewton):
```

Comment:
Pay attention to the naming convention, how about 'optimizer'?
I will discourage `naive_ctrl_pts`, do you have better name?
Different optimizers need different arguments, better to pass optimizer variables instead of optimizer class name.

Reply:
It's not possible to create the optimizer object before PnP begins, because the optimizer's constructor takes the `BetasOptimizationObjective` as input. BetasOptimizationObjective's constructor takes an intermediate solution `beta` as input. 

as shown in these two lines
```
        objective = self.BetasOptimizationObjective(solutions['beta'] * solutions['scale'].unsqueeze(-1))
        gn = self.refinement_optimizer(objective)
```